### ISSUE TITLE
Вопросы по заданию

### ISSUE THREAD

**Issue by vlas86** [systemo-biz/cp-free-busy#1]
Хотел поинтересоваться как систематически будет всё выглядеть? 
1. Кнопка добавляется в качестве виджета и там же отображает занятость персоны? 
2. Хотелось бы уточнить что именно будет указано в таксономии?

**Comment by aiiddqd** [comment]
1. шорткод, добавим мы его в пост или в виджет - это не важно
2. вопрос не понятен. там 3 термина, которые надо указывать в зависимости от нажатия.

**Comment by vlas86** [comment]
Что именно значит приватная таксономия? При смене цвета кнопки должно выводиться определённое значение в поле таксономии? Как это визуально должно выглядеть, объясните как особо одарённому.

**Comment by vlas86** [comment]
Было бы не плохо хотя бы понимать как это должно выглядеть визуально.

**Comment by aiiddqd** [comment]
пока делай без привата.
при проверке покажу.
иначе можешь не понять в чем разница

**Comment by vlas86** [comment]
Хочу уточнить, в выпадающем меню кнопки должно быть 3 цвета как я понял: зел., красный, желт., когда в меню кнопки выбираем цвет, который информирует пользователя, меняется цвет самой кнопки? Если так, то могу ли я использовать jQuery в таком случае или обойтись каким то определённым инструментарием? Есть ли макет или аналогия?

**Comment by aiiddqd** [comment]
так если смена данных должна быть через AJAX, то без jQuery никак.
эти везщи взаимоувязаны.
сменил данное - сменил цвет

**Comment by aiiddqd** [comment]
как планируешь менять цвет через jQuery? менять свойство элемента или класс?

**Comment by vlas86** [comment]
свойство элемента, примерный набросок следующий 
соответственно каждый id соответствует цвету 

window.onload = $('#menu').hide();
(function(){
$('#showMenu').on('click', function(){
$('#menu').slideToggle();
})
$('.menuItem').on('click', function(){
$('#showMenu').css('background', this.id)
})
})();

возможно не в ту степь меня потянуло, вот хочу уточнить

**Comment by aiiddqd** [comment]
Да не в ту.
Цвет меняется классом. Это же бутстрап. Не надо усложнять. Посмотри готовые классы и просто меняй их.

**Comment by vlas86** [comment]
Понял, буду пробовать

**Comment by vlas86** [comment]
Если задать в качестве id bootstrap-класс нужного цвета и при нажатии будет обнуляться класс и ему присваиваться значение id таким образом?

li id="brand-danger" class="menuItem">Red</li>
li id="brand-warning" class="menuItem">Yellow</li>
li id="brand-success" class="menuItem">Green</li>

window.onload = $("#menu").hide();
(function(){
$("#showMenu").on("click", function(){
$("#menu").slideToggle();
})
$(".menuItem").on("click", function(){
$("#showMenu").removeAttr("class");
$("#showMenu").addClass(this.id);
})
})();

**Comment by aiiddqd** [comment]
нет.
id нужен для идентификации
class для придания свойств

бесполезно пытаться менять свойства меняя id

**Comment by vlas86** [comment]
Подойдёт ли такой вариант для кнопки? 
 div class="btn-group ">
        button class="btn btn-default btn-sm dropdown-toggle " type="button" data-toggle="dropdown" >
            Small button <span class="caret"></span>

```
    /button>
  ul class="dropdown-menu" role="menu" aria-expanded="false">
   li ><a class="btn btn-danger">Red</a></li>
   li ><a class="btn btn-warning">Yellow</a></li>
   li ><a class="btn btn-success">Green</a></li>

  /ul>
/div>

<script src="http://code.jquery.com/jquery-1.10.2.js"></script>
<script src="http://netdna.bootstrapcdn.com/bootstrap/3.0.0/js/bootstrap.js"></script>
<script>

    (function(){
        var buttonClass = $('button.btn').attr('class');
        $('li a').on('click', function(){
            var liClass = $(this).attr('class');
        $('div.btn-group button').removeAttr('class');
        $('div.btn-group button').addClass(buttonClass+' '+liClass);
        })
    }());
```

</script>

**Comment by aiiddqd** [comment]
Ты пиши чтобы работало. 
А по результату уже буду смотреть код и ошибки.

**Comment by aiiddqd** [comment]
пригласил тебя в группу стажеров.
и дал доступ на этот реп.
залей тот код что уже есть

**Comment by aiiddqd** [comment]
как успехи?

**Comment by vlas86** [comment]
работаю, подключение js некоторые вопросы вызвало и проверка, постараюсь вечером залить

**Comment by vlas86** [comment]
Закомитал, но нуждаюсь в подсказке, по какой то причине перестаёт работать кнопка если подключаю файл js. Пытаюсь всё ещё разгрести, буду пробовать дальше. но если посмотрите и не испугаетесь и даже сможете подсказать что делаю не так, буду весьма признателен. И ещё хотел узнать достаточно реализации процедурной? Или же в качестве класса нужно формировать?

**Comment by aiiddqd** [comment]
foo.js убери
скрипт вставь прямо в php/html

**Comment by aiiddqd** [comment]
и дай ошибку которая пишет в консоли

**Comment by vlas86** [comment]
Закомитал с кодом внутри php-файла.
Ошибка в консоли: "Uncaught TypeError: object is not a function".

**Comment by aiiddqd** [comment]
У тебя jquery подключен?
Если нет, то ты не можешь использовать знак $ https://github.com/systemo-biz/cp-free-busy/blob/master/cp-free-busy.php#L31

Надо заменить все доллары на jQuery или подключить его.

**Comment by mbiwapa** [comment]
Приветствую, где можно смотреть примерчики? я возможно чаще буду отвечать чем толик, возможно и помогу. Толик а зачем менять знаки? Ведь скрипт и так запущен в режиме no conflict https://github.com/systemo-biz/cp-free-busy/blob/master/cp-free-busy.php#L29

**Comment by aiiddqd** [comment]
noconflict это режим работы jQuery. 
режим работы означает что jQuery должен хотя бы работать. а для этого он долен быть подключен.
если он не подключен то он не работает, а в этом случае не важен режим. нет работы - режим не поможет.

**Comment by mbiwapa** [comment]
Ну да, просто я не обратил внимания что там зависимость от jquery не указана, думал указана. достаточно указать зависимость от Jquery и WP сам подключит библиотеку.

**Comment by aiiddqd** [comment]
Это можно сделать если ты подгружаешь скрипт. А тут нет подгрузки.
Потому надо просто иницилизировать его.

**Comment by vlas86** [comment]
сделал, но всё равно такая же ошибка, единственное где не менял строку и знак $, 
https://github.com/systemo-biz/cp-free-busy/blob/master/cp-free-busy.php#L29
не подскажете в чем всё таки я ошибся, какое направление смотреть?
Что касается подключения, jQuery же подключается в header.php темы?

**Comment by mbiwapa** [comment]
нет, в шаблоне нельзя подключать, это не правильно. Правильно вот так в плагине.
function my_scripts_method() {
    wp_enqueue_script( 'jquery' );  
}    

add_action( 'wp_enqueue_scripts', 'my_scripts_method' );

**Comment by vlas86** [comment]
А по поводу работы скрипта не подскажешь? Сделал как подсказали, но результат хромает и ошибка та же. Вот голову ломаю в каком направлении смотреть. Что то видимо с анонимной ф-цией, но вот что именно?

**Comment by mbiwapa** [comment]
ну во первых, поскольку по стольку библиотеку jquery ты уже подключил, то теперь в тех строках где ты $ заменил, верни его обратно(хотя по моему там можно и jquery использовать но точно не помню), так как он у тебя был в ноу конфликте. Есть сайт пример? смотреть нужно что бы что то точнее сказать

**Comment by vlas86** [comment]
Да есть, тут заливаю и тестирую http://bestmovies.zz.mu/

**Comment by aiiddqd** [comment]
Во первых запиши скрипт по человечьи. Обычным HTML.
Закрой пхп, напиши хтмл, открой пхп.
А через ретурн - это кривой стиль написания.

**Comment by aiiddqd** [comment]
где подгрузка jquery? 
какого хера мы тут тратим на тебя время, разжовываем все тебе, кладем в рот, остается только чуть чуть жопу поднять. а результатов нет?

**Comment by mbiwapa** [comment]
Ошибка у тебя вообще не к твоему скрипту относится,там справа указанно к чему она относится, погугли как правильно присваивать и отбирать классы. Наверняка уже есть уйма готовых решений

**Comment by mbiwapa** [comment]
http://jsfiddle.net/papyby7j/7/ посмотри, сделал на примере одной красной кнопки, я в js тоже не силен, но все работает.

**Comment by vlas86** [comment]
VitoK72, Благодарю, но вроде мой код отрабатывает по сути так же. 
http://jsfiddle.net/bulvar/uepw1Lcu/ 
Скорее всего где то косяк когда прикручивал в Wordpress, буду смотреть.

**Comment by aiiddqd** [comment]
Где загрузка jquery?

**Comment by vlas86** [comment]
работаю, как по рекомендации выше закомитал, ещё дорабатываю

**Comment by aiiddqd** [comment]
зачем два раза вешать две функции на один хук? 
https://github.com/systemo-biz/cp-free-busy/blob/master/cp-free-busy.php#L48

не надо регистрировать jquery. он уже есть в ядре. надо только иницилизировать. одну строку лишнюю удали.

вот это https://github.com/systemo-biz/cp-free-busy/blob/master/cp-free-busy.php#L26

перепиши так чтобы было внутри HTML дерева самой кнопки.
то как ты сделал это хрупко потому что:
1. будет грузиться на всех страницах даже там где нет кнопки. зачем?
2. не факт что ты выбрал правильный хук. сделаешь правильно - и проблема сама отпадет.

**Comment by vlas86** [comment]
благодарю, закомитал, у меня отрабатывает нормально, единственное сомнение может я не так понял 2-ва последних пункта и соответственно недоделал

**Comment by mbiwapa** [comment]
https://github.com/systemo-biz/cp-free-busy/blob/master/cp-free-busy.php#L44  не понимаю зачем шорткод вешать на хук? Сейчас у тебя инициализация вообще висит после https://github.com/systemo-biz/cp-free-busy/blob/master/cp-free-busy.php#L41 а эта строка заканчивает выполнение функции. Все что после нее не будет выполнено. Далее вот это вот https://github.com/systemo-biz/cp-free-busy/blob/master/cp-free-busy.php#L11 просто удали и там поставь закрывающий тег php то есть "?>" (без кавычек) Далее удали 40 и 41 строку там поставь открывающий тэг <?php   И удали 44 строку, будет супер.
Видимо ты просто не понимал чего мы от тебя хотим.
Как я понял Толик предлагает тут вообще не использовать хук, а сделать как я тебе показал, то есть пустить инициализацию JQuery в генерацию страницы, тогда он подвале будет, пожалуй там ему и место, так как где нужно другим плагинам, он будет на нужном месте.

**Comment by aiiddqd** [comment]
Да уж. Ну и логика. Вырви глаз и выжги мозг называется.

Какого хера функция вывода шорткода и подключения шрифта объединены?

Вывод шорткода это одна функция.

Подключение скрипта- другая.

Включи мозг.

**Comment by mbiwapa** [comment]
Толик, я твои мысли выше правильно передал или нет?) Просто человек не очень понимает все нами сказанное, главное старается.

**Comment by aiiddqd** [comment]
Нет. Нельзя выводить jquery в нарушение кодекса.
Также как глупо вещать его вывод через шорткод.
Также как глупо писать js скрипт через инициализацию.
Парень не читает инструкции. Не задаёт вопросы. И собрал какое то чудище. Которое как то работает.

**Comment by mbiwapa** [comment]
Ну на счет js я согласен, просто оставить его в шорткоде и все. А на счет JQuery, я просто нашел в кодексе такие слова:
С версии 3.3. wp_enqueue_script() можно вызывать во время генерации страницы. В этом случае вызываемый скрипт будет подключен в подвале, в момент срабатывания события wp_footer.
Решил что оставить его в шорткоде будет самое норм, тогда jquery будет только там где кнопка, а в других местах другие плагины его сами вызовут. Видимо я не прав

**Comment by aiiddqd** [comment]
Будет. Только другие плагины об этом не будут знать. Получишь два подключения.  И кучу главной боли.

**Comment by vlas86** [comment]
yumashev, парень вопросы задает, но никто из колыбели в институт не пошёл сразу, попросил то что сделал выбросить, я выбросил, начинаю вопросы задавать ты пенишься сразу, Виктор всё по толку и без предьяв и оскорблений помогает, если чем то не устраиваю, так отвечай просто подходит решение или нет, я не экстрасенс, сказал, я делаю, на жопе ровно не сижу. Виктор и на расстоянии внимает что я не понимаю посылов, а у тебя только подачи с "пожеланиями", ты отозвался помочь и я благодарен, но не стоит поливать никого если сделал одолжение, не думаю что у тебя всё сразу ровно шло  

Закомитал версию очередную.

**Comment by aiiddqd** [comment]
Алилуйя! Задача в 50 строк кода на пол часа работы решалась 2 недели.
Я не могу нормально к такому относиться. Если делаешь лажу и не хорошие вещи - будешь читать лажу и не хорошие вещи. Ты же заставляешь меня нервничать и читать не хорошие вещи - ради справедливости ты тоже должен понервничать и почитать не хорошие вещи.
Виталя сам недавно из такой же колыбели выбрался. Ему можно. А я уже старый. Нервный.
Если тебя это успокоит, то у нас были случаи когда стажеры такую фигню писали на протяжении полу года, но все же выкорабкались и сейчас работают. А есть те кто с первых минут себе такого не позволял и начал нормально работать в первый же месяц. Мы работаем со всеми кто не сдается. Если не сдашься - значит прорвешься.

Теперь осталось самое интересное - прибить  AJAX чтобы данные сохранялись в систему.

**Comment by mbiwapa** [comment]
Так то скоро год как я выбрался хд)))))))) С аяксом думаю уже быстрее будет) Как со мной было, Задавай вопросы, если не мжешь сформулировать вопрос. Так и говори. я не могу сформулировать вопрос и хз че делать вообще))

**Comment by maksimzinchenko** [comment]
Строгий... но справедливыыый :)

**Comment by vlas86** [comment]
Добрый день! Извините пропал, у нас тут война под боком, обостряется всё, инета не было, еще и света к тому же пару дней. Хотел спросить на какой файл-обработчик ссылается ajax?

**Comment by aiiddqd** [comment]
Не очень хорошо понял вопрос, но если сделать так http://yandex.ru/yandsearch?text=ajax%20wordpress&lr=11116&suggest_reqid=223254916140576807097529984468329&csg=212%2C2394%2C14%2C6%2C0%2C0%2C0

То думаю первые же 5 пунктов снимут 99% вопросов.

**Comment by vlas86** [comment]
интересовался по поводу url-ссылки 
$.ajax({
  url: "some.php" // этого параметра
версия что сделал до вопроса http://jsfiddle.net/bulvar/uepw1Lcu/3/
на сколько я понял попросту вынести код в отдельную функцию и сделать хук под него 
add_action('wp_ajax_my_action', 'some_function'); ?
И ещё вопрос: "всё обрабатывается только в файле function.php или задействуются внешние файлы обработчики в которым идёт запрос через ajax?"

**Comment by aiiddqd** [comment]
Снова не понял вопросы. Ты не умеешь читать? Там же на русском все написано.

Ну или задай вопрос так чтобы его можно было понять.

**Comment by vlas86** [comment]
Сделал ajax, прошу проверить, понимаю, что долго, но работаю как есть возможность

**Comment by vlas86** [comment]
Ответьте хоть что-нибудь, хочу дальше продолжить но нужна консультация т.к. не знаю правильно ли делаю.

**Comment by aiiddqd** [comment]
где проверить?
перешел по ссылке и получил ошибку http://monosnap.com/image/BqtrXrvSyXkM8BLPnB1kiN2jtFLS4v

**Comment by vlas86** [comment]
пофиксил

**Comment by aiiddqd** [comment]
Пчм нет стрелки которая показывает возможность выпадания списка?

**Comment by vlas86** [comment]
подкорректирую

**Comment by vlas86** [comment]
сделал, прошу проверить!

**Comment by aiiddqd** [comment]
странный лаг. 
выбираю цвет, слово начинает отличаться от фона и это на какое то время.
думаю надо сделать чтобы цвет и текст менялся одновременно. с задержкой или без - не важно. важно чтобы одновременно.

**Comment by vlas86** [comment]
сделал, прошу проверить!

**Comment by aiiddqd** [comment]
Прям жесткие какие то проблемы.
Так сложно просто взять пример и скопировать чуть поправив?

Как так получилось что кнопки в список засунулись и список превратился в ужасный черте что?

В примере такого нет http://getbootstrap.com/components/#btn-dropdowns

Это я так понимаю удалось на ходу придумать и изуродовать меню?

**Comment by vlas86** [comment]
Бывает, ищу себе сам проблемы, помимо того что деревянный, пытаюсь выполнить, но нет того сбоку, кто подскажет, потому и раскладываюсь как могу.
А с меню такой косяк был потому что менял value() через ajax, а color используя только jquery. Потому и разница такая. 
Вариант не подходит?

**Comment by aiiddqd** [comment]
Нет конечно. Смотрится то ужасно.
Надо этим потом пользоваться, а не плеваться.
И если есть возможность сделать нормально, значит надо сделать нормально.
В чем проблема?

**Comment by vlas86** [comment]
Был косяк совершенно непонятный, я скинул на ftp несколько раз редактированный фал, и всё равно на локальном сервере всё работает нормально, на хосте - нет, и как не скидывал в итоге совершенно разные размеры получаются. Попробовал ещё раз залить, прошу проверить.

**Comment by vlas86** [comment]
Можно не проверять, после первой загрузки все было ок, теперь опять задержка, буду разбираться. Хотел без таймера обойтись.

**Comment by aiiddqd** [comment]
Мне кажется причина лишь в том, что у тебя какой то сложный jQuery.
Там по сути можно все сделать в одну или две функции.
А у тебя их куча. Вероятно ты как то не так мыслишь.

Алгоритм должне быть такой:
1. Ловим событие клика по одному из пунктов меню. Пусть там будут дата атрибуты как термины таксономии.
2. Далее по клику передаем дата атрибут в AJAX.
3. AJAX этот дата атрибут устанавливает как термин таксономии и возвращает результат
4. Далее идет обработка результата AJAX запроса и ставим тот класс что соответствует поставленному термину.

Все. 4 шага - одной функцией делаются в лет.

У тебя же этих селекторов и функций куча. Не понятно зачем так много их.

**Comment by vlas86** [comment]
да, я сейчас переделываю полностью, не годится это,наворотил лишнего
